File: solutions/binary_search.rb
def binary_search(arr, target)
  return false if arr.empty?

  middle = arr.length / 2
  middle_value = arr[middle]

  return true if middle_value == target

  search_side = middle_value > target ? arr[0...middle] : arr[middle + 1..-1]

  return binary_search(search_side, target)
end

# BONUS: MODIFY YOUR CODE TO RETURN THE INDEX OF THE TARGET, -1 IF NOT FOUND
def binary_search_index(arr, target)
  return -1 if arr.empty?

  middle = arr.length / 2
  middle_value = arr[middle]

  return middle if middle_value == target

  if middle_value > target
    return binary_search_index(arr[0...middle], target)
  else
    idx = binary_search_index(arr[middle + 1..-1], target)

    return -1 if idx == -1

    return idx + middle + 1
  end 
end

if __FILE__ == $PROGRAM_NAME
  puts "Expecting: true"
  puts "=>", binary_search([1, 2, 3], 3)

  puts

  puts "Expecting: false"
  puts "=>", binary_search([3, 5, 9], 10)

  puts

  puts "Expecting: true"
  puts "=>", binary_search([3, 5, 9, 20], 3)

  puts

  puts "Expecting: false"
  puts "=>", binary_search([3, 5, 9, 20], -10)

  puts

  puts "Expecting: true"
  puts "=>", binary_search([3, 5, 9, 20], 9)

  puts

  puts "Expecting: false"
  puts "=>", binary_search([], 9)

  # Don't forget to add your own!

  # UNCOMMENT FOR BONUS
  puts 

  puts "Expecting: 0"
  puts "=>", binary_search_index([1, 2, 3], 1)

  puts

  puts "Expecting: -1"
  puts "=>", binary_search_index([4, 7, 20], 100)

  puts

  puts "Expecting: 6"
  puts "=>", binary_search_index([1, 2, 3, 4, 5, 6, 7, 8], 7)

  puts

  puts "Expecting: 4"
  puts "=>", binary_search_index([4, 7, 20, 30, 40], 40)

  puts

  puts "Expecting: 2"
  puts "=>", binary_search_index([1, 2, 3, 4, 5, 6, 7, 8], 3)
end

# Add a written explanation of your solution

File: solutions/binary_search.js
function binarySearch(arr, target) {
  if (arr.length === 0) {
    return false;
  }

  const middle = Math.floor(arr.length / 2);
  const middleValue = arr[middle];

  if (middleValue === target) {
    return true;
  }

  const searchSide = middleValue > target ? 
    arr.slice(0, middle) : arr.slice(middle + 1);
  
  return binarySearch(searchSide, target);
}

// BONUS: MODIFY YOUR CODE TO RETURN THE INDEX OF THE TARGET, -1 IF NOT FOUND
function binarySearchIndex(arr, target) {
  if (arr.length === 0) {
    return -1;
  }

  const middle = Math.floor(arr.length / 2);
  const middleValue = arr[middle];

  if (middleValue === target) {
    return middle;
  }

  if (middleValue > target) {
    return binarySearchIndex(arr.slice(0, middle), target);
  } else {
    const idx = binarySearchIndex(arr.slice(middle + 1), target);

    if (idx === -1) {
      return -1;
    }

    return idx + middle + 1;
  }
}

if (require.main === module) {
  // add your own tests in here
  console.log("Expecting: true");
  console.log("=>", binarySearch([1, 2, 3], 3));

  console.log("");

  console.log("Expecting: false");
  console.log("=>", binarySearch([3, 5, 9], 10));

  console.log("");

  console.log("Expecting: true");
  console.log("=>", binarySearch([3, 5, 9, 20], 3));

  console.log("");

  console.log("Expecting: false");
  console.log("=>", binarySearch([3, 5, 9, 20], -10));

  console.log("");

  console.log("Expecting: true");
  console.log("=>", binarySearch([3, 5, 9, 20], 9));

  console.log("");

  console.log("Expecting: false");
  console.log("=>", binarySearch([], 9));

  // UNCOMMENT FOR BONUS
  console.log("");
  console.log("Expecting: 0");
  console.log("=>", binarySearchIndex([1, 2, 3], 1));

  console.log("");

  console.log("Expecting: -1");
  console.log("=>", binarySearchIndex([4, 7, 20], 100));

  console.log("");

  console.log("Expecting: 6");
  console.log("=>", binarySearchIndex([1, 2, 3, 4, 5, 6, 7, 8], 7));

  console.log("");

  console.log("Expecting: 4");
  console.log("=>", binarySearchIndex([4, 7, 20, 30, 40], 40));

  console.log("");

  console.log("Expecting: 2");
  console.log("=>", binarySearchIndex([1, 2, 3, 4, 5, 6, 7, 8], 3));
}

module.exports = {
  binarySearch,
  binarySearchIndex
};

// Add a written explanation of your solution
